File: Subset Sum Problem - GFG/subset-sum-problem.java
// { Driver Code Starts
import java.io.*;
import java.util.*;

class GFG
{
    public static void main(String args[])throws IOException
    {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while(t-- > 0)
        {
            int N = Integer.parseInt(read.readLine());
            String input_line[] = read.readLine().trim().split("\\s+");
            int arr[]= new int[N];
            for(int i = 0; i < N; i++)
                arr[i] = Integer.parseInt(input_line[i]);
            int sum = Integer.parseInt(read.readLine());

            Solution ob = new Solution();
            if(ob.isSubsetSum(N, arr, sum))
            System.out.println(1);
            else
            System.out.println(0);

            
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution{


    static boolean isSubsetSum(int N, int arr[], int sum){
        int[][] dp = new int[N+1][sum+1];
        for(int  []d:dp)
        {
            Arrays.fill(d,-1);
        }
        
        int w= isPossible(arr,sum,N,dp);
        
        if(w==1)
        {
            return true;
        }
        return false;
    }
    static int isPossible(int[] arr,int sum,int N,int[][] dp){
        if(sum<0)
        {
            return 0;
        }
        if(sum == 0){
            return dp[N][sum]= 1;
        }
        if(N == 0){
            return  dp[N][sum] = 0;
        }
        if(dp[N][sum]!= -1){
            return dp[N][sum];
        }
        int taken = 0;
        if(arr[N-1]<=sum){
          taken = isPossible(arr,sum- arr[N-1],N-1,dp);
        }
        int notTaken = isPossible(arr,sum,N-1,dp);
        if(taken ==1 || notTaken ==1)
        {
        return  dp[N][sum]=1;  
        }
        
        return dp[N][sum] = 0;
    }
    
}

File: Subset Sum Problem - GFG/subset-sum-problem.cpp
// { Driver Code Starts
//Initial template for C++

#include<bits/stdc++.h> 
using namespace std; 

 // } Driver Code Ends
//User function template for C++

class Solution{   
public:
    bool helper(vector<int>&arr, int currIdx, int currSum, int targetSum, vector<vector<int>>&dp){
        if(currSum == targetSum)    return true;
        if(currSum > targetSum) return false;
        if(currIdx==arr.size()) return false;
        
        // string key = currIdx+"_"+currSum;
        // if(dp.find(key)!=dp.end())  return dp[key];
        if(dp[currIdx][currSum]!=-1)    return dp[currIdx][currSum];
        
        bool includeRes = helper(arr, currIdx+1, currSum+arr[currIdx], targetSum, dp);
        bool excludeRes = helper(arr, currIdx+1, currSum, targetSum, dp);
        
        return dp[currIdx][currSum] = includeRes || excludeRes;
    }

    bool isSubsetSum(vector<int>arr, int sum){
        vector<vector<int>> dp(arr.size()+1,vector<int>(sum+1,-1));
        return helper(arr, 0, 0, sum, dp);
    }
};

// { Driver Code Starts.
int main() 
{ 
    int t;
    cin>>t;
    while(t--)
    {
        int N, sum;
        cin >> N;
        vector<int> arr(N);
        for(int i = 0; i < N; i++){
            cin >> arr[i];
        }
        cin >> sum;
        
        Solution ob;
        cout << ob.isSubsetSum(arr, sum) << endl;
    }
    return 0; 
} 
  // } Driver Code Ends